dali/pipeline/executor/executor.h

```diff
@@ -94,13 +97,23 @@ class DLL_PUBLIC Executor : public ExecutorBase, public WorkspacePolicy, public
         exec_error_(false),
         queue_sizes_(prefetch_queue_depth),
         mixed_op_stream_(0),
-        gpu_op_stream_(0) {
+        gpu_op_stream_(0),
+        get_memory_stats_(get_memory_stats) {
     DALI_ENFORCE(batch_size_ > 0, "Batch size must be greater than 0.");
     DALI_ENFORCE(device_id >= 0, "Device id must be non-negative.");
 
     stage_queue_depths_ = QueuePolicy::GetQueueSizes(prefetch_queue_depth);
   }
 
+  ~Executor() {
+    if (get_memory_stats_) {
+      std::cout << "Operator memory statistics" << std::endl;
+      PrintStats(cpu_memory_stats_);
+      PrintStats(mixed_memory_stats_);
+      PrintStats(gpu_memory_stats_);
+    }
+  }
```

Comment:
How about querying the numbers? You can return a dictionary of those numbers to Python (albeit not to TF), similarly to the Reader stats or the testing thingy that we're adding for hw decoder.
Maybe it's worth to provide a generic mechanism for returning such dictionary with metadata about any type of statistics per operator and for whole pipeline?

Reply:
Also with several Pipelines in Python you're not sure which one is which, with multi process approach like horovod you probably can mix the outputs of different Pipelines.

Alternatively some function that allows you to force print the statistics could probably help a bit.

Reply:
> Maybe it's worth to provide a generic mechanism for returning such dictionary with metadata about any type of statistics per operator and for whole pipeline?

I would go for the pipeline as operators are not a visible entities from the python side.

Reply:
The advantage of having this in the destructor is lack of synchronization. If you can query it on your own you need to introduce reader/writer lock.

Reply:
> The advantage of having this in the destructor is lack of synchronization. If you can query it on your own you need to introduce reader/writer lock.

Yes, and a TF support without any hassle. This could be an option for the TF plugin, to dump the statistics at some point when destructing the pipeline.

> I would go for the pipeline as operators are not a visible entities from the python side.

Accessible through Pipeline yes, I just meant some dictionaries like:
{ "op_name" : <statistics> } where <statics> can be another dictionary so you have easy mapping from std:: to python types.
The contents should be dependent on some options, that way you will be able to have one entry point that would be extendable.

Reply:
And we probably would want some stats about the whole pipeline as well at some point.

Reply:
Done